base/abstractarray.jl

```diff
@@ -583,7 +583,7 @@ zero{T}(x::AbstractArray{T}) = fill!(similar(x), zero(T))
 # own, LinearSlow's CartesianRange is more complicated and requires explicit
 # inlining.
 start(A::AbstractArray) = (@_inline_meta(); itr = eachindex(A); (itr, start(itr)))
-next(A::AbstractArray,i) = (@_inline_meta(); (idx, s) = next(i[1], i[2]); (A[idx], (i[1], s)))
+next(A::AbstractArray,i) = (@_inline_meta(); (idx, s) = next(i[1], i[2]); @inbounds ret = (A[idx], (i[1], s)); ret)
```

Comment:
We hadn't done this before because the old `@inbounds` for `Array` propagates through inlined functions, meaning that `@inbounds next(::Array, …)` worked as intended.  In general, `next` is only safe if you properly follow the iteration protocol… which is something only the caller knows.  So I don't think we should always annotate this with `@inbounds`.

Now that we have the second user-extensible `@inbounds` mechanism, I think we should just mark this function as `@propagate_inbounds`.  That'll allow `@inbounds next(…)` to just work for all array types, as well as `@inbounds for x in A`.